tests/e2e/attention/generate_e2e_fa2_tests.py

```diff
@@ -0,0 +1,516 @@
+#!/usr/bin/env python3
+# Copyright 2024 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+"""Generator for e2e fa2 tests.
+"""
+
+import argparse
+import enum
+import dataclasses
+import typing
+import math
+
+
+# Data type of kernel entries. The string values must match MLIR data types.
+@enum.unique
+class QueryElemTypeId(enum.Enum):
+    NONE = ""
+    F16 = "f16"
+
+
+# Data type of input entries. The string values must match MLIR data types.
+@enum.unique
+class KeyElemTypeId(enum.Enum):
+    NONE = ""
+    F16 = "f16"
+
+
+# Data type of input entries. The string values must match MLIR data types.
+@enum.unique
+class ValueElemTypeId(enum.Enum):
+    NONE = ""
+    F16 = "f16"
+
+
+# Data type of input entries. The string values must match MLIR data types.
+@enum.unique
+class ResultElemTypeId(enum.Enum):
+    NONE = ""
+    F16 = "f16"
+
+
+# Enumerates of the collections of shapes that we can generate tests for.
+# The values are the accepted values for the --shapes= flag.
+@enum.unique
+class ShapesId(enum.Enum):
+    SMALL = "small"
+    MEDIUM = "medium"
+    LARGE = "large"
+
+
+# Enumerates ways to construct MLIR tensor types.
+@enum.unique
+class Dynamicity(enum.Enum):
+    STATIC = "static"  # Use fixed values everywhere. Example: tensor<4x6xf16>.
+
+
+# Enumerates ways to initialize input buffer contents.
+@enum.unique
+class InputGenerator(enum.Enum):
+    ZERO = "zero"  # Fill with zeros
+    RANDOM = "random"  # Fill with (deterministic) pseudorandom values.
+
+
+# Enumerates ways to initialize kernel buffer contents.
+@enum.unique
+class KernelGenerator(enum.Enum):
+    ZERO = "zero"  # Fill with zeros
+    RANDOM = "random"  # Fill with (deterministic) pseudorandom values.
+
+
+@dataclasses.dataclass
+class TestShapeAndScale:
+    batch: int
+    m: int
+    k1: int
+    k2: int
```

Comment:
I suggest using more familiar names like batch, head_dim, head_count, and seq_len.

Reply:
Not a big fan of those terms personally. batch, m, n, k1, k2 are much simpler, if we can add some documentation for them.

They are essentially:

batch: obvious
m: M dimension of first and second matmul
n: N dimension of second matmul
k1: K dimension of first matmul
k2: K dimension of second matmul